### PR TITLE
HPCC-17425 Various fixes for running HPCC in windows linux subsystem

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browers
  - [ ] The component(s) render as expected

## Testing:

Tested system builds and can run queries within the windows linux subsystem.  (Still problems with Thor).

### CODE REVIEW

**Comment by ghalliday** [comment]
@mckellyln please review.  I haven't yet run tests on native linux or OSX.  It is possible the noexecstack option may cause problems.  Assuming the compiles suppport it I don't think our code requires it - we don't use trampolines.

**Comment by richardkchapman** [comment]
Checklist is missing

**Comment by mckellyln** [comment]
I checked all our exes and shared libs on native Linux in /opt/HPCCSystems and none had the execstack bit set.

**Review by mckellyln — APPROVED** [review]
Approved

**Comment by richardkchapman** [comment]
My reading of the documentation at https://linux.die.net/man/8/execstack is that this flag should already be set correctly by gcc - so the fact that our generated workunit dlls DO seem to set it suggests that (perhaps) they need it?

**Comment by richardkchapman** [comment]
Reassigning to Gavin to investigate why the flag gets set on the generated workunits - perhaps the assembler resources?

**Comment by richardkchapman** [comment]
@ghalliday

**Comment by mckellyln** [comment]
Try adding
.section  .note.GNU-stack,"",@progbits
to asm source.

**Comment by ghalliday** [comment]
@richardkchapman please can you try this version on osx

**Comment by ghalliday** [comment]
Note if we target the ARM architecture the @ will need to change to a % !

**Comment by richardkchapman** [comment]
a.out.res.s:436:20: error: unexpected token in '.section' directive
 .section .note.GNU-stack,"",@progbits
                   ^

File: ecl/hqlcpp/hqlres.cpp [hunk 1]
```diff
@@ -477,6 +477,7 @@ bool ResourceManager::flush(StringBuffer &filename, const char *basename, bool f
         unsigned id = s.id;
         VStringBuffer binfile("%s_%s_%u.bin", filename.str(), type, id);
         VStringBuffer label("%s_%u_txt_start", type, id);
+        fprintf(f, " .section .note.GNU-stack,\"\",@progbits\n");   // Prevent the stack from being marked as executable
```
Comment (richardkchapman) [inline ecl/hqlcpp/hqlres.cpp]: Perhaps needs to be moved into the else section below?

**Comment by richardkchapman** [comment]
@ghalliday Back to you

**Comment by ghalliday** [comment]
@mckellyln do you happen to know what should be generated for osx?  And whether this an osx issue or a clang issue.

**Comment by mckellyln** [comment]
@ghalliday I think this is an OSX issue.  And by default I believe the stack on OSX is not executable so we should not have to add anything to OSX asm files.
See these comments from https://developer.apple.com/library/content/documentation/Security/Conceptual/SecureCodingGuide/Articles/BufferOverflows.html:
{noformat}
Non-Executable Stack and Heap
Recent processors support a feature called the NX bit that allows the operating system to mark certain parts of memory as non-executable. If the processor tries to execute code in any memory page marked as non-executable, the program in question crashes.

OS X and iOS take advantage of this feature by marking the stack and heap as non-executable. This makes buffer overflow attacks harder because any attack that places executable code on the stack or heap and then tries to run that code will fail.

Note: For 32-bit OS X apps, if your app allows execution on OS X prior to v10.7, only the stack is marked non-executable, not the heap.
Most of the time, this is the behavior that you want. However, in some rare situations (such as writing a just-in-time compiler), it may be necessary to modify that behavior.

There are two ways to make the stack and heap executable:

Pass the -allow_stack_execute flag to the compiler. This makes the stack (not the heap) executable.
Use the mprotect system call to mark specific memory pages as executable.
{noformat}

**Comment by ghalliday** [comment]
@richardkchapman I have moved that line, and tested on windows and power8.  Please can you retest.

**Pull request merged**